lib/scan-includes.js

```diff
@@ -0,0 +1,98 @@
+    if (!scopeUrl.endsWith('/')) {
+        throw new Error(`${JSON.stringify(scopeUrl)} isn't a good scope URL.`)
+    }
+    if (path.startsWith('/') ||
+        path.includes('../') ||
+        // Block schemes, a bit more aggressively than necessary.
+        path.includes(':')) {
+        return undefined;
+    }
+
+    let resolvedUrl;
+    try {
+        resolvedUrl = new URL(path, scopeUrl).href;
+    } catch {
+        return undefined;
+    }
+    if (!resolvedUrl.startsWith(scopeUrl)) {
+        throw new Error(`Path restrictions in resolveSubUrl(${JSON.stringify(scopeUrl)}, ${JSON.stringify(path)}) failed to catch an upward navigation.`);
+    }
+    return resolvedUrl;
+
+}
+
+/** Scans for recursively included paths from a root specification.
+ *
+ * @param {string} repositoryUrl The base URL of the repository. The function
+ * won't fetch anything outside of this scope. This must end with a `/`.
+ * @param {string} rootSpec A path relative to the `repositoryUrl` pointing to
+ * the root specification in the `specType` format.
+ * @param {"bikeshed"|"respec"} specType
+ *
+ * @returns {Promise<Set<string>>} Paths relative to `repositoryUrl` that are
+ * recursively included by the specification, not including the specification itself.
+ */
+module.exports = async function scanIncludes(repositoryUrl, rootSpec, specType, fetchUrl = realFetchUrl) {
+    if (!repositoryUrl.endsWith('/')) {
+        throw new Error(`${repositoryUrl} must end with a '/'.`);
+    }
+    let includeRE;
+    switch (specType) {
+        case "bikeshed":
+            includeRE = /^path:([^\n]+)$/gm;
+            break;
+        case "respec":
+            includeRE = /data-include=["']?([^"'>]+)["']?/g;
+            break;
+        default:
+            throw new Error(`Unexpected specification type: ${JSON.stringify(specType)}`);
+    }
+
+    /** All the paths that successfully fetched go here. This is the result of
+     * the overall scan. */
+    const recursiveIncludes = new Set();
+    /** All the files we ever fetched, for the purpose of short-circuiting. */
+    const everFetched = new Set();
+
+    /** Fetches path relative to repositoryUrl, adds it to recursiveIncludes if
+     * the fetch worked, and then recursively scans any includes found in the
+     * response.
+     */
+    async function scanOneFile(path) {
+        if (everFetched.has(path)) return;
+        everFetched.add(path);
+
+        const resolvedUrl = resolveSubUrl(repositoryUrl, path);
+        if (resolvedUrl === undefined) return;
+
+        let body;
+        try {
+            body = await fetchUrl(resolvedUrl, GITHUB_SERVICE);
+        } catch {
+            // This include scan can have false positives, so if one
+            // fails to fetch, we assume it's not a real include and
+            // just ignore the fetch failure.
+            return;
+        }
+        recursiveIncludes.add(path);
+        await Promise.all(Array.from(body.matchAll(includeRE),
```

Comment:
So I must admit I’d be slightly more confortable running those sequentially, essentially because I know nothing as to how/if pooling is managed, etc. Happy to be convinced this is safe, however. I just don’t know. (And sorry, I should have raised that earlier.)

Reply:
`request` appears to use Node's [`http.Agent`](https://nodejs.org/api/http.html#http_class_http_agent) for connection pooling, and it's hard to figure out what that does by default. Some bits of the documentation indicate that it natively groups the requests into a single connection as long as the requests overlap, but other bits, in particular `maxSockets` defaulting to `Infinity`, indicate that making a bunch of concurrent requests will open a new concurrent connection for each request. And it doesn't look like `request` or `axios` support HTTP/2 at all, which would let the server handle this efficiently. :-(

It does look possible to configure all of this by creating our own `Agent` in `fetch-url.js`, but I'd rather merge this with serial requests and then optimize later.

Reply:
> It does look possible to configure all of this by creating our own Agent in fetch-url.js, but I'd rather merge this with serial requests and then optimize later.

Yup. this is the conclusion I come to every time I (re)look into this. :D